### PR TITLE
One Create Function API call and Update Connection String Prompt

### LINKED ISSUE #18930: "Create Azure Function with SQL binding" doesn't prompt user about connection string setting
1. Open existing function project
2. Right click on a table and create a new binding

The user is walked through adding the binding but is never prompted to select the connection string setting to use or whether to save the password in the settings file.

### LINKED ISSUE #18932: Duplicate "Must have Functions Project" warnings
1. Close all folders/workspaces
2. Right click on table and select "Create new Azure Function with SQL binding"

First this error pops up

image

Then if you click "Create Azure Function Project" this pops up

image

This isn't a great experience to have essentially duplicate messages like this.

### PR DESCRIPTION
**Pull request by VasuBhog**
This PR fixes #18932 and fixes #18930.

I found that due to this prompt (when a user has no azure function project open):
image

The user would be able to create an azure function project by us calling the createFunction first to create a dummy Azure Function Project in order to get certain parameters for the SQL Input/Output Binding (which requires the binding type and project folder path).

After some discussion, we want to directly create a SQL template binding instead of the above process. As such I needed to refactor the code:
1. Find the Azure Function Project 
  a. Azure Function Found (show multiple in quick pick selection or if one found then use path)
  b. No Azure Function Found in the workspace (show a warning: 1. Create Azure Function Project 2. Learn More) isCreateNewFunction = true.

2. Prompt For Binding Type
3. Based on the entry point of the user:
  a. Command Palette:
    - Prompt for Connection Profile, Prompt for Database, Prompt for Object Name
  b. Right click on table:
    - Construct object name via connectionInfo schema and table name

4. Get connectionString via the Connection Profile 
5. Prompt for Azure Function Name (based on objectName)
6. Set TemplateID based on selectedBinding.type
7. promptAndUpdateConnectionString - (isCreateNewFunction = false, not creating a new function project)
  a. Prompts user for Connection String Setting Name 
  b. Prompts user to enter Password manually into connection string
8. Create Function API - settings {
    langauge: C#
    targetFramework: netcoreapp3.1,
    templateID - selectedBinding.type,
    functionName,
    folderPath - project file folder path,
    SQL template settings {
        connectionStringSettingName,
        selectedBinding type, 
        object (if input binding), 
        table (if output binding)
      }
    }

9. if (isCreateNewFunction) 
  a. we are now able to get the projectFile Path and therefore able to set the local.settings.json with the connectionString and prompt the user to enter the password manually into the connection string
 
[Design prototype](https://www.figma.com/file/3f9jwshitDGh79bFNkiFHV/Create-Azure-Function-with-SQL-Binding-Design?node-id=6%3A846)

### CODE REVIEW

File: extensions/sql-bindings/src/common/azureFunctionsUtils.ts [hunk 1]
```diff
@@ -246,10 +246,11 @@ export async function addNugetReferenceToProjectFile(selectedProjectFile: string
  * Adds the Sql Connection String to the local.settings.json
  * @param connectionString of the SQL Server connection that was chosen by the user
  */
-export async function addConnectionStringToConfig(connectionString: string, projectFile: string): Promise<void> {
+export async function addConnectionStringToConfig(connectionString: string, projectFile: string, settingName?: string): Promise<void> {
 	const settingsFile = await getSettingsFile(projectFile);
+	let connectionStringSettingName = settingName ? settingName : constants.sqlConnectionStringSetting;
```
Comment (Charles-Gagnon) [inline extensions/sql-bindings/src/common/azureFunctionsUtils.ts:251]: You can also just set the default in the signature `settingName: string = constants.sqlConnectionStringSetting`

File: extensions/sql-bindings/src/common/azureFunctionsUtils.ts [hunk 2]
```diff
@@ -424,106 +427,111 @@ export async function promptAndUpdateConnectionStringSetting(projectUri: vscode.
...
-						connectionString = await vscode.window.showInputBox(
-							{
-								title: constants.enterConnectionString,
-								ignoreFocusOut: true,
-								value: 'Server=localhost;Initial Catalog={db_name};User ID=sa;Password={your_password};Persist Security Info=False',
-								validateInput: input => input ? undefined : constants.valueMustNotBeEmpty
+					try {
+						const projectFolder: string = path.dirname(projectUri.fsPath);
+						if (existingConnectionString) {
+							// if user is creating azure function with sql binding then we already have connection string and just need to prompt user for password
```
Comment (Charles-Gagnon) [inline extensions/sql-bindings/src/common/azureFunctionsUtils.ts:435]: You don't need to specify how a user might already have a connection string - that just makes it annoying to update if we add more ways that do this in the future. And it isn't really relevant to what this logic is doing anyways.

File: extensions/sql-bindings/src/common/azureFunctionsUtils.ts [hunk 3]
```diff
@@ -545,3 +553,22 @@ export async function promptAndUpdateConnectionStringSetting(projectUri: vscode.
 	}
 	return connectionStringSettingName;
 }
+
+export async function promptConnectionStringPassword(connectionString: string, connectionInfo?: IConnectionInfo): Promise<string> {
+	// prompt user for entering in password or not
+	// Ask user to enter password if auth type is sql login and password is not saved
```
Comment (Charles-Gagnon) [inline extensions/sql-bindings/src/common/azureFunctionsUtils.ts:559]: You never check if the password is already saved/available

File: extensions/sql-bindings/src/common/azureFunctionsUtils.ts [hunk 4]
```diff
@@ -545,3 +553,22 @@ export async function promptAndUpdateConnectionStringSetting(projectUri: vscode.
 	}
 	return connectionStringSettingName;
 }
+
+export async function promptConnectionStringPassword(connectionString: string, connectionInfo?: IConnectionInfo): Promise<string> {
```
Comment (Charles-Gagnon) [inline extensions/sql-bindings/src/common/azureFunctionsUtils.ts:557]: The fact that this is called `promptConnectionStringPassword` but then returns a full connection string is confusing. I'd suggest something more like `fillInConnectionStringPasswordIfNeeded`

File: extensions/sql-bindings/src/common/azureFunctionsUtils.ts [hunk 5]
```diff
@@ -545,3 +553,22 @@ export async function promptAndUpdateConnectionStringSetting(projectUri: vscode.
...
+		let userPassword = await vscode.window.showInputBox({
+			prompt: constants.enterPasswordPrompt,
+			placeHolder: constants.enterPasswordManually,
+			ignoreFocusOut: true,
+			password: true,
+			validateInput: input => input ? undefined : constants.valueMustNotBeEmpty
+		});
+		if (userPassword) {
+			// if user enters password replace password placeholder with user entered password
+			connectionString = connectionString.replace(constants.passwordPlaceholder, userPassword);
```
Comment (Charles-Gagnon) [inline extensions/sql-bindings/src/common/azureFunctionsUtils.ts:570]: What if the password doesn't have a placeholder at all - or it's a different type of placeholder? We should be handling those scenarios too best we can.

File: extensions/sql-bindings/src/services/azureFunctionsService.ts [hunk 6]
```diff
@@ -20,33 +20,92 @@ export async function createAzureFunction(node?: ITreeNodeInfo): Promise<void> {
...
-			TelemetryReporter.createActionEvent(TelemetryViews.CreateAzureFunctionWithSqlBinding, TelemetryActions.startCreateAzureFunctionWithSqlBinding)
+	let isCreateNewProject: boolean = false;
+	let newFunctionFileObject: azureFunctionsUtils.IFileFunctionObject;
+
+	try {
+		const azureFunctionApi = await azureFunctionsUtils.getAzureFunctionsExtensionApi();
+		if (!azureFunctionApi) {
+			propertyBag.exitReason = exitReason;
+			TelemetryReporter.createErrorEvent(TelemetryViews.CreateAzureFunctionWithSqlBinding, TelemetryActions.exitCreateAzureFunctionQuickpick)
+				// .withConnectionInfo(node?.connectionInfo)
```
Comment (Charles-Gagnon) [inline extensions/sql-bindings/src/services/azureFunctionsService.ts:35]: Remove if not needed

**Review by Charles-Gagnon — COMMENTED** [review]
You have a failing test as well

**Comment by VasuBhog** [comment]
1st Scenario (user has no azure function project folder):
user clicks `Create Azure Function Project`
image

user chooses binding type
image

user enters azure function name
image

user enters namespace
image

due to issue (https://github.com/microsoft/azuredatastudio/issues/18780) we have to set the azureFunctionWebJobStorage and in doing so we remove the prompt in the issue but get an overwrite message the user would have to press yes:
image

user asked to include password for connection string
image

if yes to above - we finish creating project and adding connection string to local.settings.json
if no - user shown to enter password:
image
- if user enters valid password then we add it to the connection string 
- if escape we show warning:
   image

**Comment by VasuBhog** [comment]
2nd Scenario (user has an azure function project):

user chooses binding type
image

user enters azure function name
image

user selects existing connection string setting or creates new
image

If user clicks create new setting name:
image

user asked to include password for connection string
image

if yes to above - we finish creating project and adding connection string to local.settings.json
if no - user shown to enter password:
image
- if user enters valid password then we add it to the connection string 
- if escape we show warning:
   image

user enters namespace
image

creates the azure function with SQL binding

**Comment by VasuBhog** [comment]
Converting to draft as this will go in after the release.

**Pull request closed without merging**